src/device/emulated/ringbuf.rs

```diff
@@ -0,0 +1,122 @@
+pub(crate) const RINGBUF_PAGE_SIZE: usize = 4096;
+pub(crate) const RINGBUF_DEPTH: usize = 128;
+pub(crate) const RINGBUF_ELEMENT_SIZE: usize = 32;
+
+#[derive(Clone, Copy)]
+pub(crate) struct RingbufPointer<const D: usize> {
+    idx_with_guard: usize,
+}
+
+const fn is_power_of_2(v: usize) -> bool {
+    (v & (v - 1)) == 0
+}
+
+impl<const D: usize> RingbufPointer<D> {
+    const GUARD_MASK: usize = D;
+    const IDX_MASK: usize = D - 1;
+
+    fn new(init_value: usize) -> Self {
+        if !is_power_of_2(D) {
+            panic!("invalid ringbuf depth.");
+        }
+        Self {
+            idx_with_guard: init_value,
+        }
+    }
+
+    fn get_idx(&self) -> usize {
+        self.idx_with_guard & Self::IDX_MASK
+    }
+    fn get_guard(&self) -> usize {
+        self.idx_with_guard & Self::GUARD_MASK
+    }
+    pub fn get_index_with_guard(&self) -> usize {
+        self.idx_with_guard
+    }
+
+    fn incr_pointer(&mut self, v: usize) {
+        // Note: should check v not greater than ringbuf depth here, but it will take extra cpu effort.
+        self.idx_with_guard += v;
+    }
+}
+
+pub(crate) struct Ringbuf<const D: usize, const E: usize> {
+    buffer: Vec<u8>,
+    start_pa: usize,
+    align_offset: usize,
+
+    head: RingbufPointer<D>,
+    tail: RingbufPointer<D>,
+}
+
+impl<const D: usize, const E: usize> Ringbuf<D, E> {
+    pub(crate) fn new() -> Self {
+        let ringbuf_size_in_byte = D * E;
+        if ringbuf_size_in_byte < RINGBUF_PAGE_SIZE {
+            panic!("invalid ringbuf size.");
```

Comment:
Please be careful with `panic`, if it's a recoverable error, use `Error` instead.